File: ios/RecipeClipperTests/Fakes/FakeMealPlanRepository.swift
import Combine
import Foundation
@testable import RecipeClipper

/// A fake that models the plan rather than only recording calls (Android's
/// FakeMealPlanRepository): `meals` and `types` are real state, `observeDays` filters and orders
/// them as the SQL does, and deleting a user type moves its meals to Dinner.
final class FakeMealPlanRepository: MealPlanRepository {
    static let breakfast: Int64 = 1
    static let lunch: Int64 = 2
    static let dinner: Int64 = 3
    static let snack: Int64 = 4

    static let defaultTypes = [
        MealType(id: breakfast, name: "Breakfast", builtInKey: "breakfast", sortOrder: 0),
        MealType(id: lunch, name: "Lunch", builtInKey: "lunch", sortOrder: 1),
        MealType(id: dinner, name: "Dinner", builtInKey: MealType.dinner, sortOrder: 2),
        MealType(id: snack, name: "Snack", builtInKey: "snack", sortOrder: 3),
    ]

    let types = CurrentValueSubject<[MealType], Never>(defaultTypes)
    let meals = CurrentValueSubject<[PlannedMeal], Never>([])
    /// recipeId → title, for meals added through `addRecipe`.
    var titles: [Int64: String] = [:]

    private var nextId: Int64 = 1000
    private var deleted: [Int64: PlannedMeal] = [:]

    func observeMealTypes() -> AnyPublisher<[MealType], Never> { types.eraseToAnyPublisher() }

    func observeDays(start: Int64, end: Int64) -> AnyPublisher<[PlannedMeal], Never> {
        meals.combineLatest(types)
            .map { all, order in
                let rank = Dictionary(uniqueKeysWithValues: order.enumerated().map { ($1.id, $0) })
                return all.enumerated()
                    .filter { $0.element.day >= start && $0.element.day <= end }
                    .sorted {
                        ($0.element.day, rank[$0.element.mealTypeId] ?? .max, $0.offset)
                            < ($1.element.day, rank[$1.element.mealTypeId] ?? .max, $1.offset)
                    }
                    .map(\.element)
            }
            .eraseToAnyPublisher()
    }

    func addRecipe(recipeId: Int64, day: Int64, mealTypeId: Int64, servings: Int?) async {
        nextId += 1
        meals.value.append(PlannedMeal(
            id: nextId, day: day, mealTypeId: mealTypeId, recipeId: recipeId,
            title: titles[recipeId] ?? "Recipe \(recipeId)", imageUrl: nil, servings: servings, note: nil
        ))
    }

    func addNote(_ note: String, day: Int64, mealTypeId: Int64) async {
        let text = note.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { return }
        nextId += 1
        meals.value.append(PlannedMeal(
            id: nextId, day: day, mealTypeId: mealTypeId, recipeId: nil,
            title: nil, imageUrl: nil, servings: nil, note: text
        ))
    }

    func move(entryId: Int64, day: Int64, mealTypeId: Int64) async {
        guard let meal = meals.value.first(where: { $0.id == entryId }) else { return }
        meals.value.removeAll { $0.id == entryId }
        meals.value.append(PlannedMeal(
            id: meal.id, day: day, mealTypeId: mealTypeId, recipeId: meal.recipeId,
            title: meal.title, imageUrl: meal.imageUrl, servings: meal.servings, note: meal.note
        ))
    }

    func delete(entryId: Int64) async -> DeletedMeal? {
        guard let meal = meals.value.first(where: { $0.id == entryId }) else { return nil }
        meals.value.removeAll { $0.id == entryId }
        deleted[meal.id] = meal
        return DeletedMeal(entry: MealPlanEntryRecord(
            id: meal.id, day: meal.day, mealTypeId: meal.mealTypeId, recipeId: meal.recipeId,
            servings: meal.servings, note: meal.note, sortOrder: 0, updatedAt: 0
        ))
    }

    func restore(_ deleted: DeletedMeal) async {
        guard let meal = self.deleted.removeValue(forKey: deleted.entry.id) else { return }
        meals.value.append(meal)
    }

    func addMealType(name: String) async {
        let text = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { return }
        nextId += 1
        types.value.append(MealType(id: nextId, name: text, builtInKey: nil, sortOrder: types.value.count))
    }

    func renameMealType(id: Int64, name: String) async {
        types.value = types.value.map {
            $0.id == id ? MealType(id: $0.id, name: name, builtInKey: $0.builtInKey, sortOrder: $0.sortOrder) : $0
        }
    }

    func reorderMealTypes(_ orderedIds: [Int64]) async {
        let byId = Dictionary(uniqueKeysWithValues: types.value.map { ($0.id, $0) })
        types.value = orderedIds.enumerated().compactMap { index, id in
            byId[id].map { MealType(id: $0.id, name: $0.name, builtInKey: $0.builtInKey, sortOrder: index) }
        }
    }

    func deleteMealType(id: Int64) async {
        guard let type = types.value.first(where: { $0.id == id }), !type.isBuiltIn else { return }
        meals.value = meals.value.map {
            $0.mealTypeId == id
                ? PlannedMeal(id: $0.id, day: $0.day, mealTypeId: Self.dinner, recipeId: $0.recipeId,
                              title: $0.title, imageUrl: $0.imageUrl, servings: $0.servings, note: $0.note)
                : $0
        }
        types.value.removeAll { $0.id == id }
    }
}

/// A pinned today and first day of the week. Defaults: Wednesday 2026-09-23, weeks from Monday.
final class FakePlanCalendar: PlanCalendar {
    /// 2026-09-23, a Wednesday.
    static let wednesday: Int64 = 20_719
    static let mondayFirst = 2
    static let sundayFirst = 1

    var todayValue: Int64
    var firstDay: Int

    init(today: Int64 = wednesday, firstDayOfWeek: Int = mondayFirst) {
        todayValue = today
        firstDay = firstDayOfWeek
    }

    func today() -> Int64 { todayValue }
    func firstDayOfWeek() -> Int { firstDay }
}

File: app/src/test/java/com/example/recipeclipper/fake/FakeMealPlanRepository.kt
package com.example.recipeclipper.fake

import com.example.recipeclipper.data.MealPlanRepository
import com.example.recipeclipper.data.local.entity.MealPlanEntryEntity
import com.example.recipeclipper.data.model.MealType
import com.example.recipeclipper.data.model.PlannedMeal
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.combine

/**
 * A fake that models the plan rather than only recording calls: [meals] and [types] are real
 * state, [observeDays] filters and orders them as the SQL does (day, then meal type order,
 * then insertion), and deleting a type moves its meals to Dinner, as the DAO does. Recipe
 * titles for new meals come from [titles].
 */
class FakeMealPlanRepository : MealPlanRepository {

    val types = MutableStateFlow(DEFAULT_TYPES)
    val meals = MutableStateFlow<List<PlannedMeal>>(emptyList())

    /** recipeId → title, for meals added through [addRecipe]. */
    val titles = mutableMapOf<Long, String>()

    private var nextId = 1000L

    override fun observeMealTypes(): Flow<List<MealType>> = types

    override fun observeDays(start: Long, end: Long): Flow<List<PlannedMeal>> =
        combine(meals, types) { all, order ->
            val rank = order.withIndex().associate { (i, t) -> t.id to i }
            all.filter { it.day in start..end }
                .withIndex()
                .sortedWith(compareBy({ it.value.day }, { rank[it.value.mealTypeId] ?: Int.MAX_VALUE }, { it.index }))
                .map { it.value }
        }

    override suspend fun addRecipe(recipeId: Long, day: Long, mealTypeId: Long, servings: Int?) {
        meals.value = meals.value + PlannedMeal(
            id = nextId++, day = day, mealTypeId = mealTypeId, recipeId = recipeId,
            title = titles[recipeId] ?: "Recipe $recipeId", imageUrl = null, servings = servings, note = null
        )
    }

    override suspend fun addNote(note: String, day: Long, mealTypeId: Long) {
        if (note.isBlank()) return
        meals.value = meals.value + PlannedMeal(
            id = nextId++, day = day, mealTypeId = mealTypeId, recipeId = null,
            title = null, imageUrl = null, servings = null, note = note.trim()
        )
    }

    override suspend fun move(entryId: Long, day: Long, mealTypeId: Long) {
        val meal = meals.value.firstOrNull { it.id == entryId } ?: return
        meals.value = meals.value.filterNot { it.id == entryId } + meal.copy(day = day, mealTypeId = mealTypeId)
    }

    override suspend fun delete(entryId: Long): MealPlanRepository.DeletedMeal? {
        val meal = meals.value.firstOrNull { it.id == entryId } ?: return null
        meals.value = meals.value.filterNot { it.id == entryId }
        deleted[meal.id] = meal
        return MealPlanRepository.DeletedMeal(
            MealPlanEntryEntity(
                id = meal.id, day = meal.day, mealTypeId = meal.mealTypeId, recipeId = meal.recipeId,
                servings = meal.servings, note = meal.note, sortOrder = 0, updatedAt = 0
            )
        )
    }

    private val deleted = mutableMapOf<Long, PlannedMeal>()

    override suspend fun restore(deleted: MealPlanRepository.DeletedMeal) {
        val meal = this.deleted.remove(deleted.entity.id) ?: return
        meals.value = meals.value + meal
    }

    override suspend fun addMealType(name: String) {
        if (name.isBlank()) return
        types.value = types.value + MealType(nextId++, name.trim(), builtInKey = null, sortOrder = types.value.size)
    }

    override suspend fun renameMealType(id: Long, name: String) {
        types.value = types.value.map { if (it.id == id) it.copy(name = name.trim()) else it }
    }

    override suspend fun reorderMealTypes(orderedIds: List<Long>) {
        val byId = types.value.associateBy { it.id }
        types.value = orderedIds.mapIndexedNotNull { i, id -> byId[id]?.copy(sortOrder = i) }
    }

    override suspend fun deleteMealType(id: Long) {
        val type = types.value.firstOrNull { it.id == id } ?: return
        if (type.isBuiltIn) return
        val dinner = types.value.first { it.builtInKey == MealType.DINNER }.id
        meals.value = meals.value.map { if (it.mealTypeId == id) it.copy(mealTypeId = dinner) else it }
        types.value = types.value.filterNot { it.id == id }
    }

    companion object {
        const val BREAKFAST = 1L
        const val LUNCH = 2L
        const val DINNER = 3L
        const val SNACK = 4L

        val DEFAULT_TYPES = listOf(
            MealType(BREAKFAST, "Breakfast", "breakfast", 0),
            MealType(LUNCH, "Lunch", "lunch", 1),
            MealType(DINNER, "Dinner", MealType.DINNER, 2),
            MealType(SNACK, "Snack", "snack", 3)
        )
    }
}
